Missing Jackson module for handling Java 8 date/time types

I am not sure if the dependency bumps in #920 caused the failure, but in one of our repositories `zally` (see [CDP build](https://sunrise.zalando.net/cdp/ghe/broadway/macao/l-2jip6j34b9xybq6s7rbz3au9qr/steps/0/logs)) started failing this morning with an error, saying that `Java 8 date/time type `java.time.OffsetDateTime` not supported by default`.

So either the dependency `com.fasterxml.jackson.datatype:jackson-datatype-jsr310` was accidentally removed or the module is for whatever reason not registered with Jackson anymore.

Log excerpt:

```
2021-07-08 07:27:51.053 ERROR 9 --- [nio-8080-exec-7] o.z.p.spring.web.advice.AdviceTrait      : Internal Server Error

java.lang.RuntimeException: check invocation failed: id=176 title=Use Problem JSON checkName=validate reason=java.lang.IllegalArgumentException: Java 8 date/time type `java.time.OffsetDateTi
me` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling (through reference chain: io.swagger.v3.oas.models.media.Schema["properti
es"]->java.util.LinkedHashMap["images"]->io.swagger.v3.oas.models.media.ObjectSchema["properties"]->java.util.LinkedHashMap["header"]->io.swagger.v3.oas.models.media.Schema["properties"]->ja
va.util.LinkedHashMap["created_at"]->io.swagger.v3.oas.models.media.DateTimeSchema["example"])
        at org.zalando.zally.core.RulesValidator.invoke(RulesValidator.kt:63) ~[zally-core-2.0.1-SNAPSHOT.jar!/:na]
        at org.zalando.zally.core.RulesValidator.validate(RulesValidator.kt:40) ~[zally-core-2.0.1-SNAPSHOT.jar!/:na]
        at org.zalando.zally.core.CompositeRulesValidator.validate(CompositeRulesValidator.kt:9) ~[zally-core-2.0.1-SNAPSHOT.jar!/:na]
        at org.zalando.zally.apireview.ApiViolationsController.validate(ApiViolationsController.kt:47) ~[classes!/:na]
...
Caused by: java.lang.reflect.InvocationTargetException: null
        at jdk.internal.reflect.GeneratedMethodAccessor512.invoke(Unknown Source) ~[na:na]
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
        at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
        at org.zalando.zally.core.RulesValidator.invoke(RulesValidator.kt:61) ~[zally-core-2.0.1-SNAPSHOT.jar!/:na]
        ... 94 common frames omitted
Caused by: java.lang.IllegalArgumentException: Java 8 date/time type `java.time.OffsetDateTime` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" 
to enable handling (through reference chain: io.swagger.v3.oas.models.media.Schema["properties"]->java.util.LinkedHashMap["images"]->io.swagger.v3.oas.models.media.ObjectSchema["properties"]
->java.util.LinkedHashMap["header"]->io.swagger.v3.oas.models.media.Schema["properties"]->java.util.LinkedHashMap["created_at"]->io.swagger.v3.oas.models.media.DateTimeSchema["example"])
        at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4314) ~[jackson-databind-2.12.2.jar!/:2.12.2]
        at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:4245) ~[jackson-databind-2.12.2.jar!/:2.12.2]
        at org.zalando.zally.ruleset.zalando.UseProblemJsonRule.testForProblemSchema(UseProblemJsonRule.kt:68) ~[zally-ruleset-zalando-2.0.1-SNAPSHOT.jar!/:na]
        at org.zalando.zally.ruleset.zalando.UseProblemJsonRule.validate(UseProblemJsonRule.kt:46) ~[zally-ruleset-zalando-2.0.1-SNAPSHOT.jar!/:na]
        ... 98 common frames omitted
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Java 8 date/time type `java.time.OffsetDateTime` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling (through reference chain: io.swagger.v3.oas.models.media.Schema["properties"]->java.util.LinkedHashMap["images"]->io.swagger.v3.oas.models.media.ObjectSchema["properties"]->java.util.LinkedHashMap["header"]->io.swagger.v3.oas.models.media.Schema["properties"]->java.util.LinkedHashMap["created_at"]->io.swagger.v3.oas.models.media.DateTimeSchema["example"])
        at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77) ~[jackson-databind-2.12.2.jar!/:2.12.2]
        at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1276) ~[jackson-databind-2.12.2.jar!/:2.12.2]
...
```
```

--- a/server/build.gradle.kts
+++ b/server/build.gradle.kts
@@ -146,13 +146,14 @@ subprojects {
     }
 
     dependencies {
+        implementation(platform("com.fasterxml.jackson:jackson-bom:2.12.2"))
         implementation("org.jetbrains.kotlin:kotlin-stdlib")
 
         // We define this here so all subprojects use the same version of jackson
-        implementation("com.fasterxml.jackson.module:jackson-module-parameter-names:2.12.2")
-        implementation("com.fasterxml.jackson.datatype:jackson-datatype-jsr310:2.12.2")
-        implementation("com.fasterxml.jackson.datatype:jackson-datatype-jdk8:2.12.2")
-        implementation("com.fasterxml.jackson.module:jackson-module-kotlin:2.12.2")
+        implementation("com.fasterxml.jackson.module:jackson-module-parameter-names")
+        implementation("com.fasterxml.jackson.datatype:jackson-datatype-jsr310")
+        implementation("com.fasterxml.jackson.datatype:jackson-datatype-jdk8")
+        implementation("com.fasterxml.jackson.module:jackson-module-kotlin")
         implementation("org.yaml:snakeyaml:1.24")
 
         testImplementation("com.jayway.jsonpath:json-path-assert:2.4.0")

--- a/server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/UseProblemJsonRule.kt
+++ b/server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/UseProblemJsonRule.kt
@@ -3,7 +3,9 @@ package org.zalando.zally.ruleset.zalando
 import com.fasterxml.jackson.annotation.JsonInclude
 import com.fasterxml.jackson.databind.JsonNode
 import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
 import com.google.common.io.Resources
+import io.swagger.v3.oas.models.responses.ApiResponse
 import org.zalando.zally.core.EMPTY_JSON_POINTER
 import org.zalando.zally.core.JsonSchemaValidator
 import org.zalando.zally.core.ObjectTreeReader
@@ -13,7 +15,6 @@ import org.zalando.zally.rule.api.Context
 import org.zalando.zally.rule.api.Rule
 import org.zalando.zally.rule.api.Severity
 import org.zalando.zally.rule.api.Violation
-import io.swagger.v3.oas.models.responses.ApiResponse
 
 @Rule(
     ruleSet = ZalandoRuleSet::class,
@@ -26,7 +27,11 @@ class UseProblemJsonRule {
     private val description = "Operations should return problem JSON when any problem occurs during processing " +
         "whether caused by client or server."
 
-    private val objectMapper by lazy { ObjectMapper().setSerializationInclusion(JsonInclude.Include.NON_NULL) }
+    private val objectMapper by lazy {
+        ObjectMapper()
+            .setSerializationInclusion(JsonInclude.Include.NON_NULL)
+            .registerModule(JavaTimeModule())
+    }
 
     private val problemSchemaValidator by lazy {
         val schemaUrl = Resources.getResource("schemas/problem-meta-schema.json")

--- a/server/zally-ruleset-zalando/src/test/kotlin/org/zalando/zally/ruleset/zalando/UseProblemJsonRuleTest.kt
+++ b/server/zally-ruleset-zalando/src/test/kotlin/org/zalando/zally/ruleset/zalando/UseProblemJsonRuleTest.kt
@@ -59,10 +59,6 @@ class UseProblemJsonRuleTest {
                           schema:
                             type: object
                             properties:
-                              created:
-                                type: string
-                                format: date-time
-                                example: '2020-05-14T14:22:01Z'
                               status:
                                 type: string
         """.trimIndent()

--- a/server/zally-ruleset-zalando/src/test/kotlin/org/zalando/zally/ruleset/zalando/UseProblemJsonRuleTest.kt
+++ b/server/zally-ruleset-zalando/src/test/kotlin/org/zalando/zally/ruleset/zalando/UseProblemJsonRuleTest.kt
@@ -59,6 +59,10 @@ class UseProblemJsonRuleTest {
                           schema:
                             type: object
                             properties:
+                              created:
+                                type: string
+                                format: date-time
+                                example: '2020-05-14T14:22:01Z'
                               status:
                                 type: string
         """.trimIndent()
@@ -262,5 +266,38 @@ class UseProblemJsonRuleTest {
                 "/components/schemas/Problem/properties/instance",
                 "/components/schemas/Problem/properties/instance/type"
             )
+    }
+
+    @Test
+    fun `should return no violation if date and time properties are used in the Problem object schema`() {
+        @Language("YAML")
+        val content = """
+            openapi: 3.0.1
+            info:
+              version: 1.0.0
+              title: Pets API
+            paths:
+              /bad:
+                get:
+                  responses:
+                    default:
+                      description: Lorem Ipsum
+                      content:
+                        application/problem+json:
+                          schema:
+                            type: object
+                            allOf: 
+                            - "$\ref": https://zalando.github.io/problem/schema.yaml#/Problem
+                            - type: object
+                              properties:
+                                created:
+                                  type: string
+                                  format: date-time
+                                  example: "2020-05-14T14:22:01Z" 
+        """.trimIndent()
+
+        val context = DefaultContextFactory().getOpenApiContext(content)
+        val violations = rule.validate(context)
+        ZallyAssertions.assertThat(violations).isEmpty()
     }
 }

